tempesta_fw/sock.c

```diff
@@ -1087,6 +1084,41 @@ ss_getpeername(struct sock *sk, struct sockaddr *uaddr, int *uaddr_len)
 }
 EXPORT_SYMBOL(ss_getpeername);
 
+static inline void
+__ss_do_send(SsWork *sw)
+{
+	struct sock *sk = sw->sk;
+
+	bh_lock_sock(sk);
+	ss_do_send(sk, &sw->skb_list, sw->flags);
+	if (!(sw->flags & SS_F_CONN_CLOSE)) {
+		bh_unlock_sock(sk);
+		return;
+	}
+	ss_do_close(sk);
+	bh_unlock_sock(sk);
+	SS_CALL(connection_drop, sk);
+	sock_put(sk); /* paired with ss_do_close() */
+}
+
+static inline void
+__ss_do_close(SsWork *sw)
+{
+	struct sock *sk = sw->sk;
+
+	bh_lock_sock(sk);
+	if (!ss_sock_live(sk)) {
+		SS_DBG("[%d]: %s: Socket inactive: sk %p\n",
+		       smp_processor_id(), __func__, sk);
+		bh_unlock_sock(sk);
+		return;
+	}
+	ss_do_close(sk);
+	bh_unlock_sock(sk);
+	SS_CALL(connection_drop, sk);
+	sock_put(sk); /* paired with ss_do_close() */
+}
```

Comment:
The functions are called only once from `ss_tx_action()`. Do we really need them?